contracts/cosmwasm-vm/cw-xcall/src/send_call_message.rs

```diff
@@ -17,47 +22,84 @@ impl<'a> CwCallService<'a> {
+                msg_type: MessageType::MessageWithRollback,
+                data,
+                rollback: rollback.unwrap(),
+            })
+        } else {
+            AnyMessage::CallMessage(CallMessage {
+                data,
+                msg_type: MessageType::BasicMessage,
+            })
         };
+        let envelope = Envelope::new(msg, sources, destinations);
+        self.send_call(deps, info, to, envelope)
+    }
 
-        self.ensure_rollback_length(&rollback_data)?;
-        println!("{LOG_PREFIX} Packet Validated");
+    pub fn validate_payload(
+        &self,
+        deps: Deps,
+        caller: &Addr,
+        envelope: &Envelope,
+    ) -> Result<(), ContractError> {
+        match &envelope.message {
+            AnyMessage::CallMessage(_m) => Ok(()),
+            AnyMessage::CallMessageWithRollback(m) => {
+                if !is_contract(deps.querier, caller) {
+                    return Err(ContractError::RollbackNotPossible);
+                }
+                self.ensure_rollback_length(&m.rollback().unwrap())?;
+                Ok(())
+            }
+        }
+    }
+
+    pub fn send_call(
+        &self,
+        deps: DepsMut,
+        info: MessageInfo,
+        to: NetworkAddress,
+        envelope: Envelope,
+    ) -> Result<Response, ContractError> {
+        let caller = info.sender.clone();
+        let config = self.get_config(deps.as_ref().storage)?;
+        let nid = config.network_id;
+        self.validate_payload(deps.as_ref(), &caller, &envelope)?;
 
         let sequence_no = self.get_next_sn(deps.storage)?;
-        let mut confirmed_sources = sources.clone();
+        let mut confirmed_sources = envelope.sources.clone();
         let from = NetworkAddress::new(&nid, caller.as_ref());
 
         if confirmed_sources.is_empty() {
             let default = self.get_default_connection(deps.as_ref().storage, to.nid())?;
             confirmed_sources = vec![default.to_string()]
         }
 
-        if need_response {
-            let request =
-                CallRequest::new(caller.clone(), to.clone(), sources, rollback_data, false);
+        if envelope.message.should_persist() {
+            let rollback_data = envelope.message.rollback().unwrap();
+            let request = CallRequest::new(
+                caller.clone(),
+                to.clone(),
+                envelope.sources,
+                rollback_data,
+                false,
+            );
 
             self.store_call_request(deps.storage, sequence_no, &request)?;
         }
-
         let call_request = CSMessageRequest::new(
             from,
             to.account(),
             sequence_no,
-            need_response,
-            data.to_vec(),
-            destinations,
+            envelope.message.msg_type().clone(),
+            envelope.message.data(),
+            envelope.destinations,
         );
 
+        let need_response = call_request.need_response();
```

Comment:
need_response() this would be fine but this method might need more additions as we add more messages with responses. Would this still be more clear. Since a new message would require you to edit the callrequest file

Reply:
ya thats a valid concern since we dont want to modify storage object but the flag need_response is just copy from the actual message type so unless we have aditional flags in message flow itself that needs to be handled in execution we wont have any problem and when we have such flags we would need to modify storage anyway.